### PR TITLE
Makefile,*: use tailscale.com/cmd/mkversion

### PR DESCRIPTION
**Pull request by raggi**
We've suffered misalignment in versioning and toolchain usage due to the
shell invocations downstream of ./version/tailscale-version.sh, but also
the whole version data scheme in the Makefile was quite complicated, and
required synchronization in the build.grade.

- Makfile no longer needs to be version aware itself.
- A Makefile target tailscale.version refreshes a local cached output
  from tailscale.com/cmd/mkversion which is updated when go.mod / go.sum
  change.
- build.gradle loads tailscale.version to get the version string.
- ldflags are produced from tailscale.version via version-ldflags.sh

Updates https://github.com/tailscale/tailscale/issues/13850

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -38,3 +38,5 @@ tailscale.jks
 libtailscale.aar
 libtailscale-sources.jar
 .DS_Store
+
+tailscale.version
```
Comment (bradfitz) [inline .gitignore:42]: I thought we were going to check this file in to the repo? Or did you not `git add` it?

File: Makefile [hunk 2]
```diff
@@ -141,6 +127,12 @@ tailscale-test.apk: gradle-dependencies
 	(cd android && ./gradlew assembleApplicationTestAndroidTest)
 	install -C ./android/build/outputs/apk/androidTest/applicationTest/android-applicationTest-androidTest.apk $@
 
+tailscale.version: go.mod go.sum
+    $(shell ./tool/go run tailscale.com/cmd/mkversion > tailscale.version)
+
```
Comment (bradfitz) [inline Makefile:132]: oh, so we're still doing the full git clone pain, just hidden behind this tool

**Review by bradfitz — COMMENTED** [review]
> Updates

You can use the F-Droid woes bug: https://github.com/tailscale/tailscale/issues/13850

File: Makefile (hunk 2 above)
Comment (raggi) [reply Makefile:132]: yeah, at least we now are down to one implementation of that behavior though!

File: Makefile [hunk 3]
```diff
@@ -141,6 +127,12 @@ tailscale-test.apk: gradle-dependencies
 	(cd android && ./gradlew assembleApplicationTestAndroidTest)
 	install -C ./android/build/outputs/apk/androidTest/applicationTest/android-applicationTest-androidTest.apk $@
 
+tailscale.version: go.mod go.sum .git/HEAD
```
Comment (bradfitz) [inline Makefile:130]: does having a git file as a dep here cause problems for people who download a release tarball from github without the `.git` directory?

File: Makefile (hunk 2 above)
Comment (bradfitz) [reply Makefile:132]: hah. lolsob.

File: android/build.gradle [hunk 4]
```diff
@@ -38,7 +38,7 @@ android {
         minSdkVersion 26
         targetSdkVersion 34
         versionCode 241
-        versionName "1.77.12-ta8f9c0d6e-g753b8d3fb4b"
+        versionName getVersionProperty("VERSION_LONG")
```
Comment (bradfitz) [inline android/build.gradle:41]: ooooh hell yeah

File: android/build.gradle [hunk 5]
```diff
@@ -181,3 +181,13 @@ def getLocalProperty(key, defaultValue) {
...
+
+
+def getVersionProperty(key) {
+    // tailscale.version is created / updated by the makefile, it is in a loosely
+    // Makfile/envfile format, which is also loosely a properties file format.
+    //      make tailscale.version
+    def versionProps = new Properties()
+    versionProps.load(project.file('../tailscale.version').newDataInputStream())
+    return versionProps.getProperty(key).replaceAll('^\"|\"$', '')
+}
```
Comment (bradfitz) [inline android/build.gradle:193]: groovy!

**Review by bradfitz — APPROVED** [review]

File: Makefile (hunk 3 above)
Comment (andrew-d) [reply Makefile:130]: I believe you can make this optional with `$(wildcard .git/HEAD)`, which will expand to the empty string if the file is not present. e.g.:
```
tailscale.version: go.mod go.sum $(wildcard .git/HEAD)
```

File: Makefile [hunk 6]
```diff
@@ -141,6 +127,12 @@ tailscale-test.apk: gradle-dependencies
...
+    $(shell ./tool/go run tailscale.com/cmd/mkversion > tailscale.version)
+
+version: tailscale.version ## print the current version information
```
Comment (andrew-d) [inline Makefile:133]: Slightly nitpick-y, but... make this a `PHONY` target?

**Pull request merged**